enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,545 @@
+
+### Bundle/BundleDeployment
+
+A "BundleDeployment" is an OLM API that's responsible for managing OLM's first-class Bundle API, which represents content that needs to be made available to other consumers on the cluster. These APIs are analogous to the Deployment/PodSpec resources seen throughout core Kubernetes. The BundleDeployment API allows users to specify a Bundle template, and that template contains the desired specification of a Bundle resource. The management of a Bundle and BundleDeployment resources is delegated to individual "provisioner" implementations.
+
+### Provisioner
+
+A "provisioner" is a controller that understands the Bundle and BundleDeployment APIs. Individual provisioner implementations use ingress-style "class" name identifiers to distinguish between the BundleDeployments and Bundles resources that it needs to manage. Each provisioner implementation requires the definition of a unique provisioner IDs.
+This provisioner pattern helps abstract the variations of bundle formats that are defined throughout the Kubernetes landscape. Future provisioner implementations may manage helm charts, crossplane packages, and OLM bundle formats.
+
+Both the Bundle and BundleDeployment APIs require a unique provisioner ID to be specified. This enables higher level systems (e.g. a PO manager) to generate BundleDeployments that referencing different unique provisioner IDs. This is useful for components that want more control over how to source, and apply, Bundle contents.
+
+## Proposal
+
+### Introducing a "PlatformOperator" CRD
+
+This first phase will introduce a new, top-level "PlatformOperator" cluster-scoped CRD. While admins today could include OLM manifests before cluster installation, this approach lacks the ability to influence the cluster's lifecycle. This new CRD will bring CVO awareness of these manifests, and help pave the way for adding a first-class UX for configuring POs that will be added in future phases.
+
+Given this is an openshift-only API, this proposal recommends introducing the CRD into the openshift/api repository under the `platform.openshift.io/v1alpha1` API group. During phase 0, this resource will contain an empty spec. Future phase requirements may necessitate more granular admin controls over how upgrades are approved, and how PO packages are sourced. Before cluster creation, an admin would be responsible for creating, and editing, the cluster's Kubernetes manifests with their desired PlatformOperator(s) configuration, and a FeatureGate manifest that enables the "TechPreviewNoUpgrades" feature set. Enabling POs through this FeatureGate manifest will be detailed below. The installer would then bootstrap the cluster using those custom manifests.
```

Comment:
> If there's no spec, how do I as an end user, tell the cluster what it should be installing for my PO? I can see down below that it is based on the metadata.name. 

Yep, the current proposal outlines an end user configuring the PO's metadata.Name in order to express intent to install a PO. There's some obvious UX gaps there, like "how can I find the correct package name upfront", but these are limitations we're comfortable during phase 0 while we work towards a first-class UX in subsequent phases.

> Do the restrictions imposed on this field match that of the OLM catalog? Can every catalog item name be described in this way?

That's a good question.. I think the answer is generally yes, we can use the PO resource's name to find package names with an OLM catalog. I don't think we need any regex validation here, and it's unclear whether we can even do that with a resource's metadata.Name vs. an explicit spec field.

The only potential edge case is package names that use "_" as a separator, and IIRC, that's an invalid resource name character, so we may need some internal translation of `s/_/-/g` if we don't have that kind of static validation already present when manipulating OLM's catalog contents.

Reply:
> Yep, the current proposal outlines an end user configuring the PO's metadata.Name in order to express intent to install a PO. There's some obvious UX gaps there, like "how can I find the correct package name upfront", but these are limitations we're comfortable during phase 0 while we work towards a first-class UX in subsequent phases.

Adding this limitation up front concerns me a little, as, typically when this kind of thing happens, we are reluctant to change or remodel the API later (speaking from reviewing other API evolutions), so, I'm hesitant to say we shouldn't try and think about this ahead of time rather than leaving it til later

> The only potential edge case is package names that use "_" as a separator, and IIRC, that's an invalid resource name character, so we may need some internal translation of s/_/-/g if we don't have that kind of static validation already present when manipulating OLM's catalog contents.

This is an incredibly good argument for not using the metadata.name and having a dedicated spec field imo

Reply:
I'm -1 on doing some internal translation of `metadata.name` to package name. I think the only choices are:
1. Package names _must_ follow the same schema as `metadata.name`
2. Package names are defined in a PO `.spec` field.

Reply:
Yeah, I'd agree with trying to avoid doing any translation here, especially since we have already have ways to workaround this. I think using the resource's name has nice properties, and being able to fire off list queries to get the POs that are installed in the cluster is an added bonus.

At the same time: it would be nice to limit the number of OLM, and especially runtime, changes during this first phase. But the FBC schema isn't tied to any Kubernetes API so we don't have the same guidelines there.

Reply:
Is this where we're consolidating the naming discussion?

re: dealing with duplicate names if we use `spec.packageName`, it seems like we would need a mutex _somewhere_ to guarantee avoidance of duplicates. It seems like even a validating webhook would need a mutex because you have to consider the possibility of two requests arriving simultaneously, where both have their their package names checked against the cluster prior to either being admitted.

Reply:
@joelanford It looks like we're consolidating with the original thread on this topic: https://github.com/openshift/enhancements/pull/1170#discussion_r935815024. Do you want to copy/paste your reply there?